### PR TITLE
Hotfix/suit-10520

### PR DESCRIPTION
**Pull request by honeydripper**
Sorry for the monster pull request, but it was necessary.

- Refactored API output log to make it more userfriendly
- Fixed bug with unit tests not running when .suitestrc is present
- Fixed writing logs to files on Windows
- Removed obsolete logging
- Changed automation mode to output logs instantly
- Added distinction between assertion and evaluation to the logs
- Fixed log files overwriting one another

### CODE REVIEW

File: lib/utils/getDeviceInfo.js [hunk 1]
```diff
@@ -25,19 +25,28 @@ async function getDevicesDetails(devicesIds) {
...
  */
-function getDeviceName(deviceInfo) {
-	if (deviceInfo.manufacturer || deviceInfo.model) {
-		return `${deviceInfo.manufacturer} - ${deviceInfo.model}`;
-	}
+function getDeviceName(deviceInfo, short = false) {
+	const {manufacturer, model, customName} = deviceInfo;
+
+	if (customName)
+		return short? customName : `${customName} (${manufacturer} ${model})`;
```
Comment (JeKachan) [inline lib/utils/getDeviceInfo.js]: missed space before question mark

File: lib/utils/getDeviceInfo.js [hunk 2]
```diff
@@ -25,19 +25,28 @@ async function getDevicesDetails(devicesIds) {
...
 
-	return '';
+	return short? `${model}` : `${manufacturer} ${model}`;
```
Comment (JeKachan) [inline lib/utils/getDeviceInfo.js]: the same as above

File: lib/utils/logger.js [hunk 3]
```diff
@@ -90,42 +90,123 @@ const methods = {
...
+ * @param {object} deviceDetails
+ */
+function devicePrefix(deviceDetails) {
+	if (!deviceDetails || !deviceDetails.shortDisplayName)
+		return 'Launcher ';
+
+	return deviceDetails.shortDisplayName + ' ';
+}
+
+function logLine(msg, logMethod, fontColor, leftRail = '') {
```
Comment (JeKachan) [inline lib/utils/logger.js]: Please add JsDocs to logLine and extractOpType

File: lib/utils/logger.js [hunk 4]
```diff
@@ -90,42 +90,123 @@ const methods = {
...
+		opType,
+		logMessages,
+	};
+}
+
+let lastLogSource = '';
+
+/**
+ *
+ * @param consoleMethod
```
Comment (JeKachan) [inline lib/utils/logger.js]: Please provide types for function arguments and short description for them.

File: lib/utils/opType.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
+const stripAnsiChars = require('./stringUtils').stripAnsiChars;
+
+// Our chains can be either "A" assert or "E" - evaluation.
```
Comment (givehug) [inline lib/utils/opType.js]: What about 'Query' lines ?

File: lib/utils/logger.js [hunk 6]
```diff
@@ -90,42 +90,123 @@ const methods = {
...
+ * @param logMethod
+ * @returns {*[]}
  */
-const log = method => (message, deviceId) => {
-	const currentTimestamp = config.timestamp.toLowerCase() === timestampNone
-		? ''
-		: `[${timestamp(config.timestamp)}] `;
+function outputLogsInLauncherMode(consoleMethod, messages, logMethod) {
+	const fontColor = consoleMethod.launcherColor || stColors.suit;
+	const deviceDetails = messages[messages.length - 1];
```
Comment (givehug) [inline lib/utils/logger.js]: I think `const log = method => (message, deviceId) => {` is much clearer interface

File: lib/utils/logger.js [hunk 7]
```diff
@@ -90,42 +90,123 @@ const methods = {
 
 Object.freeze(methods);
 
+function timestamp() {
+	return moment().format('MMM D HH:MM:SS');
```
Comment (givehug) [inline lib/utils/logger.js]: timestamp was configurable with cli args or rc config before

File: lib/utils/socketChainHelper.js [hunk 8]
```diff
@@ -101,7 +101,7 @@ const getRequestType = (data, hasQuery = true) => {
 };
 
 const DEFAULT_COUNT = 1;
-const DEFAULT_DELAY = 1000;
+const DEFAULT_DELAY = 1;
```
Comment (givehug) [inline lib/utils/socketChainHelper.js]: 1ms ? Is it supported by BE?

File: lib/utils/opType.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+module.exports = {
+
+	getOpType: function(data) {
+		return `|${data && data.isAssert ? 'A' : 'E'}|`;
+	},
+
+	isOpType: function(logString) {
+		return logString.match(/^\|[AE]\|$/);
+	},
+
```
Comment (givehug) [inline lib/utils/opType.js]: Would prefer to see jsdocs annotations

File: test/testLauncher/SuitestLauncher.test.js [hunk 10]
```diff
@@ -14,28 +14,28 @@ const envVars = require('../../lib/constants/enviroment');
 const logger = require('../../lib/utils/logger');
 
 describe('SuitestLauncher', () => {
-	before(async() => {
+	before(async () => {
```
Comment (givehug) [inline test/testLauncher/SuitestLauncher.test.js]: Why these changes, we don't have this enforce din eslint?

File: lib/commands/endTest.js [hunk 11]
```diff
@@ -16,15 +16,16 @@ const logger = require('../utils/logger');
...
+	const info = logger.delayed(commandWillBeExecuted('endTest'));
 	// authorize
 	const authedContent = await authContext.authorizeWs({type: wsContentTypes.endTest}, endTest.name);
 
 	// make ws request
 	await webSockets.send(authedContent);
 
 	testContext.clear();
-	logger.info(commandExecuted('endTest'));
+	clearInterval(info);
```
[the remark this answers is not in the document]
Comment (honeydripper) [reply lib/commands/endTest.js]: they are not actually immediate. `logger.delayed()` shows the log line only if there is a longer period of log silence, so that user is comforted that something is happening but it takes time. If another log would come before the timeout is reached the delayed log line should not be displayed anymore. 

However you are right about the weird `clearTimeout` which should have been an implementation detail of the logger, i'll fix that

File: lib/commands/pairDevice.js [hunk 12]
```diff
@@ -31,7 +37,9 @@ async function pairDevice(deviceId) {
 
 	pairedDeviceContext.clear();
 	pairedDeviceContext.setContext({deviceId});
```
[the remark this answers is not in the document]
Comment (honeydripper) [reply lib/commands/pairDevice.js]: deviceDetails where introduced before this pull request, I just tried to utilize them in the log. I prefer not to touch this, but feel free to update this PR as necessary if you think it is required.

File: lib/commands/pairDevice.js [hunk 13]
```diff
@@ -11,16 +11,22 @@ const {authContext, pairedDeviceContext} = require('../context');
...
  * Pair with device
  * @param {string} deviceId - device to connect to
  * @returns {ChainablePromise.<DeviceData>}
  */
 async function pairDevice(deviceId) {
-	logger.info(connectingToDevice(deviceId));
 	// validate deviceId string to be in uuid format
 	validate(validators.UUID, deviceId, invalidInputMessage(pairDevice.name, 'Device id'));
+
+	const [deviceDetails] = await getDevicesDetails([deviceId]);
```
[the remark this answers is not in the document]
Comment (honeydripper) [reply lib/commands/pairDevice.js]: The launcher needs to know the details about which devices will be used in the current run, hence this line. Why do you think this is a problem?

File: lib/commands/pairDevice.js [hunk 14]
```diff
@@ -11,16 +11,22 @@ const {authContext, pairedDeviceContext} = require('../context');
...
+	const deviceName = getDeviceName(deviceDetails);
```
[the remark this answers is not in the document]
Comment (honeydripper) [reply lib/commands/pairDevice.js]: yup, will fix that

File: lib/utils/logger.js (hunk 6 above)
Comment (honeydripper) [reply lib/utils/logger.js]: I think so too, however this is how it is implemented in master currently, I only refactored this a bit due to CodeClimate complains. 
The problem is that `logger` is used for both the launcher and the child process. And when the Launcher captures logs from the child process this is what the logger gets: the captured message and device information from the process.

File: lib/utils/logger.js (hunk 7 above)
Comment (honeydripper) [reply lib/utils/logger.js]: Ok, will fix that

File: lib/utils/socketChainHelper.js (hunk 8 above)
Comment (honeydripper) [reply lib/utils/socketChainHelper.js]: This weird 1s delay was introduced to account for slower TVs a few years ago.
This was a mistake. Changing this constant could have been in a separate PR however.

File: test/testLauncher/SuitestLauncher.test.js (hunk 10 above)
Comment (honeydripper) [reply test/testLauncher/SuitestLauncher.test.js]: my editor did this automatically. 
If you prefer i can revert this. 
It would be great if we have all IDE settings in this project somehow synchronized.

File: lib/utils/logger.js [hunk 15]
```diff
@@ -90,42 +90,123 @@ const methods = {
...
+}
+
+function logLine(msg, logMethod, fontColor, leftRail = '') {
+	let obj = msg;
+
+	if (typeof msg === 'string') {
+		try {
+			obj = JSON.parse(msg);
+		} catch (e) {
+			// something to make ESLINT happy
```
[the remark this answers is not in the document]
Comment (honeydripper) [reply lib/utils/logger.js]: no actually, it is only an attempt to see if we should treat msg as JS object and display it accrodingly.

File: lib/commands/pairDevice.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (givehug) [reply lib/commands/pairDevice.js]: Thats fine. getDevicesDetails uses authContext.authorizeHttp internally. It protects unauthorized user from making this request.

File: test/testLauncher/SuitestLauncher.test.js (hunk 10 above)
Comment (givehug) [reply test/testLauncher/SuitestLauncher.test.js]: The rule is actually present in eslint: `"space-before-function-paren": ["warn", "never"]`, space should not be there

File: lib/utils/opType.js (hunk 5 above)
Comment (honeydripper) [reply lib/utils/opType.js]: Anything that does not throw an exception when the server responds with queryFailed is considered evaluation, everything else is considered assertion. I think this covers query lines?

File: lib/chains/elementChain.js [hunk 16]
```diff
@@ -79,13 +80,18 @@ const toString = data => {
...
+				out = (data.isNegated ? elementNotMatchJS : elementMatchJS)(elName, data.comparator.val);
 				break;
-			default:
-				out = elementMatchProps(elName);
+			default: {
+				const matches = data.comparator.props.map(
+					one => `  ${ELEMENT_PROP[one.name]} ${PROP_COMPARATOR_M[one.type]} ${VALUE_M[one.val] || one.val}`
+				);
+
+				out = elementMatchProps(elName, '\n' + matches.join(',\n'));
```
[the remark this answers is not in the document]
Comment (honeydripper) [reply lib/chains/elementChain.js]: i think it would be to difficult/wasteful to apply this function since it does not support item transformation

File: test/testLauncher/SuitestLauncher.test.js (hunk 10 above)
Comment (honeydripper) [reply test/testLauncher/SuitestLauncher.test.js]: right. But i have applied autoformatting (Ctrl+Alt+L) on idea and it did this. I think we should check the .editorconfig or something. I will revert this file.

File: package.json [hunk 17]
```diff
@@ -64,8 +64,11 @@
     "ajv": "^6.1.1",
```
Comment (givehug) [inline package.json]: @honeydripper If i'm not mistaken, time-stamp module is not needed anymore, is it? If not, please remove from package.json and update yarn.lock. That's all from me.

File: lib/commands/openSession.js [hunk 18]
```diff
@@ -33,7 +33,8 @@ async function openSession(authData, _sessionType) {
 	let authedRequestObject,
 		response;
 
-	logger.info(commandWillBeExecuted('openSession'));
+	logger.delayed(sessionOpen(), 4e3);
```
Comment (JeKachan) [inline lib/commands/openSession.js]: Why this message should be postponed for 4 seconds?

File: test/chains/windowChain.test.js [hunk 19]
```diff
@@ -95,13 +95,16 @@ describe('Window chain', () => {
 
 	it('should convert to string with meaningful message', () => {
 		assert.equal(window().toString(), 'Window chain');
-		assert.equal(window().sendText('text').toString(), 'Send text \'text\' to window');
 		assert.equal(
-			window().sendText('text').repeat(3).interval(4000).toString(),
-			'Send text \'text\' to window 3 times every 4000ms'
+			window().sendText("text string").toString(),
```
Comment (JeKachan) [inline test/chains/windowChain.test.js]: should be used single quotes.

**Comment by andriichumak** [comment]
@honeydripper  Backend guys prepared a hotfix on their end to deliver device custom name in the feed. Do you want to still integrate that into this PR, or should I close this one?

**Comment by honeydripper** [comment]
@andriichumak I think we should close this PR before it gets any bigger. @givehug @JeKachan  would appreciate a review of the newly added code; due to a bug in logging had to refactor the SuitestLauncher class ...

File: lib/utils/getDeviceInfo.js [hunk 20]
```diff
@@ -41,12 +41,19 @@ async function getDevicesDetails(devicesIds) {
...
  */
 function getDeviceName(deviceInfo, short = false) {
-	const {manufacturer, model, customName} = deviceInfo;
+	const {manufacturer, model, customName, deviceId} = deviceInfo;
+
+	let name;
 
 	if (customName)
-		return short ? customName : `${customName} (${manufacturer} ${model})`;
+		name = short ? customName : `${customName} (${manufacturer} ${model})`;
```
Comment (JeKachan) [inline lib/utils/getDeviceInfo.js]: As I remember if customName not specified this field will be filled with manufacturer and model by default?

File: lib/utils/socketErrorMessages.js [hunk 21]
```diff
@@ -342,6 +350,10 @@ const errorMap = {
 				return t['errorType.queryFailed.updateAlert']();
 			case 'notFunction':
 				return t['errorType.queryFailed.notFunction'](toString(chainData));
+			case 'missingSubject':
+				if (chainData.type === 'element')
+					return `Element ${JSON.stringify(chainData.selector)} was not found.`;
```
Comment (JeKachan) [inline lib/utils/socketErrorMessages.js]: You can use 'errorType.NoSuchElement' i18n string.

File: test/utils/socketErrorMessages.test.js [hunk 22]
```diff
@@ -68,7 +68,7 @@ describe('Socket error messages', () => {
 	});
 
 	it('Error message should return specific messages', () => {
-		const toString = () => 'Chain description';
+		const toString = (data, nameOnly) => nameOnly? 'Element name' : 'Chain description';
```
Comment (JeKachan) [inline test/utils/socketErrorMessages.test.js]: space missed after nameOnly.
